### ISSUE TITLE
MIDI Event Filtering

### ISSUE THREAD

**Issue by orchetect** [orchetect/MIDIKit#18]
# Proposal

While some MIDI event filtering can be done using the native Swift `.filter { }` method on `[MIDI.Event]`, more complex filtering tasks would be better as methods implemented in MIDIKit.

## Filter methods on `[MIDI.Event]`

Consistent with Swift's `.filter { }` method, these methods will produce a new `[MIDI.Event]` including only events that match the filter.

```swift
extension Collection where Element == MIDI.Event {
    public func filterChannelVoice() -> [Element]
    public func filterSystemExclusive() -> [Element]
    public func filterSystemCommon() -> [Element]
    public func filterSystemRealTime() -> [Element]
    
    public func filterChannelVoice(channel isIncluded: MIDI.UInt4) -> [Element]
    public func filterChannelVoice(channels isIncluded: [MIDI.UInt4]) -> [Element]
    public func filter(channel isIncluded: MIDI.UInt4) -> [Element]
    public func filter(channels isIncluded: [MIDI.UInt4]) -> [Element]
}
```

Inverse methods should be offered that accomplish the opposite.

Since Swift's `.drop` method on collections is not meant as an inverse to `.filter` since it drops while a predicate is true instead of dropping all matching elements in the array, a new clear nomenclature should be used. `.filterOut` may be the most succinct.

```swift
extension Collection where Element == MIDI.Event {
    public func filterOutChannelVoice() -> [Element]
    public func filterOutSystemExclusive() -> [Element]
    public func filterOutSystemCommon() -> [Element]
    public func filterOutSystemRealTime() -> [Element]
    
    public func filterOutChannelVoice(channel isExcluded: MIDI.UInt4) -> [Element]
    public func filterOutChannelVoice(channels isExcluded: [MIDI.UInt4]) -> [Element]
    public func filterOut(channel isExcluded: MIDI.UInt4) -> [Element]
    public func filterOut(channels isExcluded: [MIDI.UInt4]) -> [Element]
}
```

## `Filter` Definition Enum and `EventFilter` Object

Additionally, new types could be introduced:
- new `Filter` enum describing a filter action, mirroring all of the ones offered as `[MIDI.Event]` collection methods as described above
- new `EventFilter` class containing one or more of the filter definitions could be introduced.

An `EventFilter` instance could then optionally be attached to an object that conforms to `SendsMIDIEvents` `ReceivesMIDIEvents` either at its midiIn or midiOut node. The filter would automatically process MIDI events before midiIn receives them, or on the way out of midiOut before it returns the events.

This brings the benefit of endpoint/module filters being applied automatically, as well as being able to stack multiple filter criteria in one encapsulated mechanism.

```swift
extension MIDI {
    open class EventFilter {
        public var filters: [MIDI.Event.Filter]
        public func filter(events: [MIDI.Event]) -> [MIDI.Event] {
            // iterate through filters array and return the result
        }
    }
}

extension MIDI.Event {
    public enum Filter {
        // filter
        case filterChannelVoice
        case filterSystemExclusive
        case filterSystemCommon
        case filterSystemRealTime
        
        case filterChannelVoice(channel: MIDI.UInt4)
        case filterChannelVoice(channels: [MIDI.UInt4])
        case filter(channel: MIDI.UInt4)
        case filter(channels: [MIDI.UInt4])

        // filter out
        case filterOutChannelVoice
        case filterOutSystemExclusive
        case filterOutSystemCommon
        case filterOutSystemRealTime
        
        case filterOutChannelVoice(channel: MIDI.UInt4)
        case filterOutChannelVoice(channels: [MIDI.UInt4])
        case filterOut(channel: MIDI.UInt4)
        case filterOut(channels: [MIDI.UInt4])
    }
}
```

### Channel Voice
- [ ] Channel Voice status type filter
  - `notes` (note on / note off)
  - `polyAftertouch`
  - `cc`
  - `programChange`
  - `chanAftertouch`
  - `pitchBend`
- [ ] Note number range limiting (lowerNote...upperNote)
- [ ] CC filter

**Comment by orchetect** [comment]
Additional filter ideas:

- Note number range limiting (lowerNote...upperNote)

**Comment by orchetect** [comment]
Additional filter ideas:

## Channel Voice
- [ ] Channel Voice status type filter (note on, note off, cc, etc.)
- [ ] Note number range limiting (lowerNote...upperNote)
- [ ] CC filter

**Comment by orchetect** [comment]
A robust set of filter methods have been implemented in MIDIKit 0.1.15.
